WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Shipping Address Validation/AddressValidator.swift

```diff
@@ -0,0 +1,91 @@
+import Yosemite
+
+/// Reusable implementation of the Address validation introduced by the
+/// `ShippingLabelAddressFormViewModel`, but striped of the specifics to serve the form UI
+/// 
+class AddressValidator {
+    let siteID: Int64
+    private let stores: StoresManager
+
+    init(siteID: Int64, stores: StoresManager) {
+        self.siteID = siteID
+        self.stores = stores
+    }
+
+    func validate(address: Address, onlyLocally: Bool, onCompletion: @escaping (Result<Void, AddressValidationError>) -> Void) {
+        let convertedAddress = ShippingLabelAddress(company: address.company ?? "",
+                                                    name: address.fullName,
+                                                    phone: address.phone ?? "",
+                                                    country: address.country,
+                                                    state: address.state,
+                                                    address1: address.address1 ,
+                                                    address2: address.address2 ?? "",
+                                                    city: address.city,
+                                                    postcode: address.postcode)
+
+        let localErrors = validateAddressLocally(addressToBeValidated: convertedAddress)
+        if localErrors.isNotEmpty {
+            let localErrorMessage = localErrors.map { $0.rawValue }.joined(separator: ", ")
+            onCompletion(.failure(.local(localErrorMessage)))
+            return
+        }
+
+        if onlyLocally {
+            onCompletion(.success(()))
+            return
+        }
+
+        let addressToBeVerified = ShippingLabelAddressVerification(address: convertedAddress, type: .destination)
+        let action = ShippingLabelAction.validateAddress(siteID: siteID, address: addressToBeVerified) { (result) in
+            switch result {
+            case .success:
+                onCompletion(.success(()))
+            case .failure(let error):
+                guard let shippingLabelError = error as? ShippingLabelAddressValidationError else {
+                    return
```

Comment:
nit: if we have unknown failure, validation process and its callback will hang forever?

Reply:
Good point! I decided to fix this by allowing a nullable error, so we can still trigger a remote failure, but if it's not of the expected type, we can't acquire the failure message. Here's the fix: https://github.com/woocommerce/woocommerce-ios/pull/7587/commits/447c944b87930d32bd0f479322e1e9051a1f8a64